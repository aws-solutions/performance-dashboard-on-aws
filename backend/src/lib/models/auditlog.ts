/*
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *  SPDX-License-Identifier: Apache-2.0
 */

export enum ItemEvent {
    Create = "Create",
    Update = "Update",
    Delete = "Delete",
}

export type ModifiedProperty = {
    property: string;
    oldValue: any;
    newValue: any;
};

export interface AuditLog {
    itemId: string;
    timestamp: Date;
    userId: string;
    event: ItemEvent;
    modifiedProperties?: ModifiedProperty[];
}

// Represents a dynamodb item in the Audit Trail table
export interface AuditLogItem {
    pk: string;
    sk: string;
    type: string;
    userId: string;
    event: ItemEvent;
    modifiedProperties?: ModifiedProperty[];
}

export interface DashboardAuditLog extends AuditLog {
    version: number;
}

// Represents a dynamodb item in the Audit Trail table that
// contains attributes specific to Dashboard audit logs.
export interface DashboardAuditLogItem extends AuditLogItem {
    version: number;
}
