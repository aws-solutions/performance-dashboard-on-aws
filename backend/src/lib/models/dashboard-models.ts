export type Dashboard = {
    id: string,
    name: string,
    topicAreaId: string,
    topicAreaName: string,
    description: string,
    createdBy: string,
};

export type DashboardList = Array<Dashboard>;

export interface DashboardItem {
    pk: string,
    sk: string,
    type: string,
    dashboardName: string,
    topicAreaId: string,
    topicAreaName: string,
    description: string,
    createdBy: string,
};