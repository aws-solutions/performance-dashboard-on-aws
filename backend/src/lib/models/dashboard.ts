/*
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *  SPDX-License-Identifier: Apache-2.0
 */

import { Widget } from "./widget";

export const DASHBOARD_ITEM_TYPE = "Dashboard";

export enum DashboardState {
    Draft = "Draft",
    Published = "Published",
    Archived = "Archived",
    /**
     * @type {string}
     * @deprecated in favor of {@link Draft}
     */
    PublishPending = "PublishPending",
    Inactive = "Inactive",
}

export interface DashboardVersion {
    id: string;
    version: number;
    state: DashboardState;
    friendlyURL?: string;
}

export interface Dashboard {
    id: string;
    name: string;
    version: number;
    parentDashboardId: string;
    topicAreaId: string;
    topicAreaName: string;
    displayTableOfContents: boolean;
    /**
     * @type {any}
     * @deprecated no longer supported
     */
    tableOfContents?: any;
    description: string;
    createdBy: string;
    updatedAt: Date;
    updatedBy?: string;
    submittedBy?: string;
    publishedBy?: string;
    archivedBy?: string;
    deletedBy?: string;
    state: DashboardState;
    releaseNotes?: string;
    widgets?: Array<Widget>;
    friendlyURL?: string;
}

export type DashboardList = Array<Dashboard>;

export interface DashboardItem {
    pk: string;
    sk: string;
    type: string;
    version: number;
    parentDashboardId: string;
    dashboardName: string;
    topicAreaId: string;
    topicAreaName: string;
    displayTableOfContents: boolean;
    tableOfContents?: any;
    description: string;
    createdBy: string;
    updatedAt: string;
    updatedBy?: string;
    submittedBy?: string;
    publishedBy?: string;
    archivedBy?: string;
    deletedBy?: string;
    state: string;
    releaseNotes?: string;
    friendlyURL?: string;
}

// Public representation of a dashboard. Hides some fields
// like createdBy and state, which should not be exposed
// on public endpoints.
export interface PublicDashboard {
    id: string;
    name: string;
    topicAreaId: string;
    topicAreaName: string;
    displayTableOfContents: boolean;
    tableOfContents?: any;
    description: string;
    updatedAt: Date;
    widgets?: Array<Widget>;
    friendlyURL?: string;

    // Filled on a homepage search with dashboard content
    // that matches the search query.
    queryMatches?: Array<string>;
}
