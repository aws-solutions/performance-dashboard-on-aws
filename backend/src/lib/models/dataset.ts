/*
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *  SPDX-License-Identifier: Apache-2.0
 */

export enum SourceType {
    IngestApi = "IngestApi",
    FileUpload = "FileUpload",
}

export enum DatasetSchema {
    None = "None",
    Metrics = "Metrics",
}

export interface Dataset {
    id: string;
    createdBy: string;
    fileName: string;
    s3Key: {
        raw: string;
        json: string;
    };
    updatedAt: Date;
    sourceType: SourceType;
    schema?: DatasetSchema;
}

export type DatasetContent = Array<object>;
export type DatasetList = Array<Dataset>;

export interface DatasetItem {
    pk: string;
    sk: string;
    type: string;
    createdBy: string;
    fileName: string;
    s3Key: {
        raw: string;
        json: string;
    };
    updatedAt: string;
    sourceType: string;
    schema?: string;
}
