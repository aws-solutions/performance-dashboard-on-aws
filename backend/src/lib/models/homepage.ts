/*
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *  SPDX-License-Identifier: Apache-2.0
 */

export interface Homepage {
    title: string;
    description: string;
    updatedAt: Date;
}

export interface HomepageItem {
    pk: string;
    sk: string;
    type: string;
    updatedAt?: string;
    title: string;
    description: string;
}
