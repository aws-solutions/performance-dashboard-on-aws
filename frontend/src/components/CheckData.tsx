/*
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *  SPDX-License-Identifier: Apache-2.0
 */

import React, { useCallback, useMemo, MouseEvent } from "react";
import { ColumnDataType, CurrencyDataType, NumberDataType } from "../models";
import TickFormatter from "../services/TickFormatter";
import UtilsService from "../services/UtilsService";
import Button from "./Button";
import Dropdown from "./Dropdown";
import Table from "./Table";
import { useTranslation } from "react-i18next";
import { useWindowSize } from "../hooks";

interface Props {
    hiddenColumns: Set<string>;
    setHiddenColumns: Function;
    backStep: (event: MouseEvent<HTMLButtonElement>) => void;
    advanceStep: (event: MouseEvent<HTMLButtonElement>) => void;
    onCancel: (event: MouseEvent<HTMLButtonElement>) => void;
    data: Array<any>;
    dataTypes: Map<string, ColumnDataType>;
    setDataTypes: Function;
    numberTypes: Map<string, NumberDataType>;
    setNumberTypes: Function;
    currencyTypes: Map<string, CurrencyDataType>;
    setCurrencyTypes: Function;
    sortByColumn?: string;
    sortByDesc?: boolean;
    setSortByColumn?: Function;
    setSortByDesc?: Function;
    reset?: Function;
    widgetType: string;
}

function CheckData(props: Props) {
    const { t } = useTranslation();
    const windowSize = useWindowSize();
    const isMobile = windowSize.width <= 600;

    const handleHideFromVisualizationChange = useCallback(
        (event: React.FormEvent<HTMLInputElement>) => {
            const target = event.target as HTMLInputElement;
            const newHiddenColumns = new Set(props.hiddenColumns);
            if (target.checked) {
                newHiddenColumns.delete(target.name);
            } else {
                newHiddenColumns.add(target.name);
            }
            props.setHiddenColumns(newHiddenColumns);
        },
        [props.hiddenColumns],
    );

    const handleDataTypeChange = useCallback(
        (event: React.FormEvent<HTMLInputElement>) => {
            const target = event.target as HTMLInputElement;
            const newDataTypes = new Map(props.dataTypes);

            if (target.value === ColumnDataType.Text) {
                newDataTypes.set(target.id, ColumnDataType.Text);
            } else if (target.value === ColumnDataType.Number) {
                newDataTypes.set(target.id, ColumnDataType.Number);
            } else if (target.value === ColumnDataType.Date) {
                newDataTypes.set(target.id, ColumnDataType.Date);
            } else {
                newDataTypes.delete(target.id);
            }

            props.setDataTypes(newDataTypes);
        },
        [props.dataTypes],
    );

    const handleNumberTypeChange = useCallback(
        (event: React.FormEvent<HTMLInputElement>) => {
            const target = event.target as HTMLInputElement;
            const newNumberTypes = new Map(props.numberTypes);

            if (target.value === NumberDataType.Percentage) {
                newNumberTypes.set(target.id, NumberDataType.Percentage);
            } else if (target.value === NumberDataType.Currency) {
                newNumberTypes.set(target.id, NumberDataType.Currency);
            } else if (target.value === NumberDataType["With thousands separators"]) {
                newNumberTypes.set(target.id, NumberDataType["With thousands separators"]);
            } else {
                newNumberTypes.delete(target.id);
            }

            props.setNumberTypes(newNumberTypes);
        },
        [props.numberTypes],
    );

    const handleCurrencyTypeChange = useCallback(
        (event: React.FormEvent<HTMLInputElement>) => {
            const target = event.target as HTMLInputElement;
            const newCurrencyTypes = new Map(props.currencyTypes);

            if (target.value === CurrencyDataType["Dollar $"]) {
                newCurrencyTypes.set(target.id, CurrencyDataType["Dollar $"]);
            } else if (target.value === CurrencyDataType["Euro €"]) {
                newCurrencyTypes.set(target.id, CurrencyDataType["Euro €"]);
            } else if (target.value === CurrencyDataType["Pound £"]) {
                newCurrencyTypes.set(target.id, CurrencyDataType["Pound £"]);
            } else {
                newCurrencyTypes.delete(target.id);
            }

            props.setCurrencyTypes(newCurrencyTypes);
        },
        [props.currencyTypes],
    );

    const handleCell = (properties: any, header: any) => {
        const row = properties.row.original;
        const cellHeader = !UtilsService.isCellEmpty(row[header])
            ? row[header].toLocaleString()
            : "-";
        if (!props.dataTypes.has(header)) {
            return cellHeader;
        }
        if (props.dataTypes.get(header) === ColumnDataType.Number) {
            return typeof row[header] === "number" ? (
                TickFormatter.format(row[header], 0, false, "", "", {
                    columnName: header,
                    hidden: props.hiddenColumns.has(header),
                    dataType: ColumnDataType.Number,
                    numberType: props.numberTypes.get(header),
                    currencyType: props.currencyTypes.get(header),
                })
            ) : (
                <div className="text-secondary-vivid">{`! ${cellHeader}`}</div>
            );
        }
        if (props.dataTypes.get(header) === ColumnDataType.Date) {
            return !isNaN(Date.parse(row[header])) ? (
                row[header].toLocaleString()
            ) : (
                <div className="text-secondary-vivid">{`! ${cellHeader}`}</div>
            );
        }
        if (props.dataTypes.get(header) === ColumnDataType.Text) {
            return cellHeader;
        }
        return cellHeader;
    };

    const checkDataTableRows = useMemo(() => props.data || [], [props.data]);
    const dataColumns = props.data.length ? Object.keys(props.data[0]) : [];
    const checkDataTableColumns = useMemo(
        () =>
            dataColumns.map((header, i) => {
                return {
                    Header: () => (
                        <span className="text-center usa-checkbox margin-bottom-1">
                            <input
                                className="usa-checkbox__input"
                                id={`checkbox-header-${i}`}
                                type="checkbox"
                                aria-label={header + " " + t("CheckBox")}
                                name={header}
                                defaultChecked={!props.hiddenColumns.has(header)}
                                onChange={handleHideFromVisualizationChange}
                            />
                            <label
                                className="usa-checkbox__label"
                                htmlFor={`checkbox-header-${i}`}
                            ></label>
                        </span>
                    ),
                    id: `checkbox${header}`,
                    columns: [
                        {
                            Header: header,
                            id: header,
                            accessor: header,
                            minWidth: 150,
                            Cell: (properties: any) => {
                                return handleCell(properties, header);
                            },
                        },
                    ],
                };
            }),
        [props.data, props.dataTypes, props.numberTypes, props.currencyTypes],
    );

    return (
        <>
            <div className={`usa-hint ${isMobile ? "grid-col-12" : "grid-col-6"}`}>
                {t("CheckDataDescription", { widgetType: props.widgetType })}
            </div>

            <div className="grid-col-12 margin-top-3 font-sans-md text-bold">
                {t("IncludeInVisualization")}
            </div>

            <div className="check-data-table grid-col-12">
                <Table
                    id="items"
                    selection="none"
                    rows={checkDataTableRows}
                    initialSortAscending={props.sortByDesc !== undefined ? !props.sortByDesc : true}
                    initialSortByField={props.sortByColumn}
                    pageSize={50}
                    disablePagination={false}
                    disableBorderless={true}
                    columns={checkDataTableColumns}
                    hiddenColumns={props.hiddenColumns}
                    addNumbersColumn={true}
                    sortByColumn={props.sortByColumn}
                    sortByDesc={props.sortByDesc}
                    setSortByColumn={props.setSortByColumn}
                    setSortByDesc={props.setSortByDesc}
                    reset={props.reset}
                    keepBorderBottom
                    mobileNavigation
                    settingTable={true}
                />
            </div>
            <br />

            <div className="grid-col-12 margin-top-3 font-sans-md text-bold">
                {t("FormatColumns")}
            </div>

            <div className="grid-col-12 font-sans-md font-sans-md text-bold">
                {(props.data.length ? Object.keys(props.data[0]) : []).map((header: string) => {
                    return (
                        <div key={header} className="margin-top-4">
                            <span className="text-base">{t("CheckDataHeader", { header })}</span>

                            <div className={isMobile ? "grid-col-8" : "grid-col-4"}>
                                <Dropdown
                                    id={header}
                                    name="dataType"
                                    label={t("DataFormat")}
                                    options={[
                                        {
                                            value: "",
                                            label: t("SelectAnOption"),
                                        },
                                        {
                                            value: ColumnDataType.Text,
                                            label: t("Text"),
                                        },
                                        {
                                            value: ColumnDataType.Number,
                                            label: t("Number"),
                                        },
                                        {
                                            value: ColumnDataType.Date,
                                            label: t("Date"),
                                        },
                                    ]}
                                    value={props.dataTypes.get(header)}
                                    onChange={handleDataTypeChange}
                                    className="margin-top-2"
                                />
                            </div>

                            {props.dataTypes.get(header) === ColumnDataType.Number && (
                                <div className={isMobile ? "grid-col-8" : "grid-col-4"}>
                                    <Dropdown
                                        id={header}
                                        name="numberType"
                                        label={t("NumberFormat")}
                                        options={[
                                            {
                                                value: "",
                                                label: t("SelectAnOption"),
                                            },
                                            {
                                                value: NumberDataType.Percentage,
                                                label: t("Percentage"),
                                            },
                                            {
                                                value: NumberDataType.Currency,
                                                label: t("Currency"),
                                            },
                                            {
                                                value: NumberDataType["With thousands separators"],
                                                label: t("WithThousandsSeparators"),
                                            },
                                        ]}
                                        value={props.numberTypes.get(header)}
                                        onChange={handleNumberTypeChange}
                                        className="margin-top-3"
                                    />
                                </div>
                            )}

                            {props.dataTypes.get(header) === ColumnDataType.Number &&
                                props.numberTypes.get(header) === NumberDataType.Currency && (
                                    <div className={isMobile ? "grid-col-8" : "grid-col-4"}>
                                        <Dropdown
                                            id={header}
                                            name="currencyType"
                                            label={t("Currency")}
                                            options={[
                                                {
                                                    value: "",
                                                    label: t("SelectAnOption"),
                                                },
                                                {
                                                    value: CurrencyDataType["Dollar $"],
                                                    label: t("Dollar"),
                                                },
                                                {
                                                    value: CurrencyDataType["Euro €"],
                                                    label: t("Euro"),
                                                },
                                                {
                                                    value: CurrencyDataType["Pound £"],
                                                    label: t("Pound"),
                                                },
                                            ]}
                                            value={props.currencyTypes.get(header)}
                                            onChange={handleCurrencyTypeChange}
                                            className="margin-top-3"
                                        />
                                    </div>
                                )}
                            <br />
                            <hr />
                        </div>
                    );
                })}
            </div>

            <br />
            <Button
                variant="outline"
                type="button"
                onClick={props.backStep}
                className="margin-top-1"
            >
                {t("BackButton")}
            </Button>
            <Button
                type="button"
                onClick={props.advanceStep}
                disabled={!props.data.length}
                className="margin-top-1"
            >
                {t("ContinueButton")}
            </Button>
            <Button
                variant="unstyled"
                className="text-base-dark hover:text-base-darker active:text-base-darkest margin-top-1"
                type="button"
                onClick={props.onCancel}
            >
                {t("Cancel")}
            </Button>
        </>
    );
}

export default CheckData;
