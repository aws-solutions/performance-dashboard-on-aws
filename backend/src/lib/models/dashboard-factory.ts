import { v4 as uuidv4 } from "uuid";
import { Dashboard, DashboardItem } from "./dashboard-models";
import { User } from "./user-models";
import topicareaFactory from "./topicarea-factory";

const DASHBOARD: string = 'Dashboard';

function createNew(name: string, topicAreaId: string, topicAreaName: string, user: User): Dashboard {
  return {
    id: uuidv4(),
    name,
    topicAreaId,
    topicAreaName,
    createdBy: user.userId,
  };
}

/**
 * Converts a Dashboard to a DynamoDB item.
 */
function toItem(dashboard: Dashboard): DashboardItem {
  let item: DashboardItem = {
    pk: itemId(dashboard.id),
    sk: itemId(dashboard.id),
    type: DASHBOARD,
    dashboardName: dashboard.name,
    topicAreaName: dashboard.topicAreaName,
    topicAreaId: topicareaFactory.itemId(dashboard.topicAreaId),
    description: dashboard.description,
    overview: dashboard.overview,
    createdBy: dashboard.createdBy,
  };
  return item;
}

/**
 * Converts a DynamoDB item into a Dashboard object
 */
function fromItem(item: DashboardItem): Dashboard {
  const id = item.pk.substring(10);
  let dashboard: Dashboard = {
    id,
    name: item.dashboardName,
    topicAreaId: item.topicAreaId.substring(10),
    topicAreaName: item.topicAreaName,
    description: item.description,
    overview: item.overview,
    createdBy: item.createdBy,
  }
  return dashboard;
}

function itemId(id: string): string { return `${DASHBOARD}#${id}` }

export default {
  toItem,
  createNew,
  fromItem,
  itemId,
};
