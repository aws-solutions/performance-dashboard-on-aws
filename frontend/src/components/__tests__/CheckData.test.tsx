/*
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *  SPDX-License-Identifier: Apache-2.0
 */

import React from "react";
import { render } from "@testing-library/react";
import CheckData from "../CheckData";
import { ColumnDataType } from "../../models";

test("renders the CheckData component", async () => {
    const wrapper = render(
        <CheckData
            data={[
                {
                    id: "1",
                    name: "Banana",
                    updatedAt: "2021-11-11",
                    icon: "fa-banana",
                },
                {
                    id: "2",
                    name: "Chocolate",
                    updatedAt: "2020-11-11",
                    icon: "fa-candy-bar",
                },
                {
                    id: "3",
                    name: "Vanilla",
                    updatedAt: "2019-11-11",
                    icon: "fa-ice-cream",
                },
            ]}
            advanceStep={() => {}}
            backStep={() => {}}
            onCancel={() => {}}
            setHiddenColumns={() => {}}
            hiddenColumns={new Set<string>()}
            dataTypes={
                new Map<string, ColumnDataType>([
                    ["id", ColumnDataType.Number],
                    ["name", ColumnDataType.Text],
                    ["updatedAt", ColumnDataType.Date],
                ])
            }
            setDataTypes={() => {}}
            currencyTypes={new Map()}
            numberTypes={new Map()}
            setCurrencyTypes={() => {}}
            setNumberTypes={() => {}}
            widgetType="the chart"
        />,
    );
    expect(wrapper.container).toMatchSnapshot();
});
