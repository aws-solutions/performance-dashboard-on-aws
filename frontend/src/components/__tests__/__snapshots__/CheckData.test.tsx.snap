// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the CheckData component 1`] = `
<div>
  <div
    class="usa-hint grid-col-6"
  >
    Please make sure that the system formats your data correctly. Select columns to format as numbers, dates, or text. Also select columns to hide or show from the chart.
  </div>
  <div
    class="grid-col-12 margin-top-3 font-sans-md text-bold"
  >
    Include in visualization
  </div>
  <div
    class="check-data-table grid-col-12"
  >
    <div
      class="overflow-x-hidden overflow-y-hidden"
    >
      <table
        class="usa-table"
        id="items"
        role="table"
      >
        <thead>
          <tr
            role="row"
          >
            <th
              aria-sort="none"
              colspan="1"
              role="columnheader"
              scope="col"
              style="min-width: 0;"
            >
              <span>
                Â 
              </span>
            </th>
            <th
              aria-sort="none"
              colspan="1"
              role="columnheader"
              scope="col"
            >
              <span>
                <span
                  class="text-center usa-checkbox margin-bottom-1"
                >
                  <input
                    aria-label="id CheckBox"
                    checked=""
                    class="usa-checkbox__input"
                    id="checkbox-header-0"
                    name="id"
                    type="checkbox"
                  />
                  <label
                    class="usa-checkbox__label"
                    for="checkbox-header-0"
                  />
                </span>
              </span>
            </th>
            <th
              aria-sort="none"
              colspan="1"
              role="columnheader"
              scope="col"
            >
              <span>
                <span
                  class="text-center usa-checkbox margin-bottom-1"
                >
                  <input
                    aria-label="name CheckBox"
                    checked=""
                    class="usa-checkbox__input"
                    id="checkbox-header-1"
                    name="name"
                    type="checkbox"
                  />
                  <label
                    class="usa-checkbox__label"
                    for="checkbox-header-1"
                  />
                </span>
              </span>
            </th>
            <th
              aria-sort="none"
              colspan="1"
              role="columnheader"
              scope="col"
            >
              <span>
                <span
                  class="text-center usa-checkbox margin-bottom-1"
                >
                  <input
                    aria-label="updatedAt CheckBox"
                    checked=""
                    class="usa-checkbox__input"
                    id="checkbox-header-2"
                    name="updatedAt"
                    type="checkbox"
                  />
                  <label
                    class="usa-checkbox__label"
                    for="checkbox-header-2"
                  />
                </span>
              </span>
            </th>
            <th
              aria-sort="none"
              colspan="1"
              role="columnheader"
              scope="col"
            >
              <span>
                <span
                  class="text-center usa-checkbox margin-bottom-1"
                >
                  <input
                    aria-label="icon CheckBox"
                    checked=""
                    class="usa-checkbox__input"
                    id="checkbox-header-3"
                    name="icon"
                    type="checkbox"
                  />
                  <label
                    class="usa-checkbox__label"
                    for="checkbox-header-3"
                  />
                </span>
              </span>
            </th>
          </tr>
          <tr
            role="row"
          >
            <th
              aria-sort="none"
              colspan="1"
              role="columnheader"
              scope="col"
              style="min-width: 0;"
            >
              <span>
                <div>
                  1
                </div>
              </span>
            </th>
            <th
              aria-sort="none"
              colspan="1"
              role="columnheader"
              scope="col"
              style="min-width: 150px;"
            >
              <button
                aria-label="Toggle SortBy: id"
                class="usa-button usa-button--unstyled text-no-underline"
                style="cursor: pointer;"
                title="Toggle SortBy: id"
                type="button"
              >
                <span>
                  id
                </span>
                <svg
                  aria-hidden="true"
                  class="svg-inline--fa fa-chevron-down fa-w-14 margin-left-1"
                  data-icon="chevron-down"
                  data-prefix="fas"
                  focusable="false"
                  role="img"
                  viewBox="0 0 448 512"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M207.029 381.476L12.686 187.132c-9.373-9.373-9.373-24.569 0-33.941l22.667-22.667c9.357-9.357 24.522-9.375 33.901-.04L224 284.505l154.745-154.021c9.379-9.335 24.544-9.317 33.901.04l22.667 22.667c9.373 9.373 9.373 24.569 0 33.941L240.971 381.476c-9.373 9.372-24.569 9.372-33.942 0z"
                    fill="currentColor"
                  />
                </svg>
              </button>
            </th>
            <th
              aria-sort="none"
              colspan="1"
              role="columnheader"
              scope="col"
              style="min-width: 150px;"
            >
              <button
                aria-label="Toggle SortBy: name"
                class="usa-button usa-button--unstyled text-no-underline"
                style="cursor: pointer;"
                title="Toggle SortBy: name"
                type="button"
              >
                <span>
                  name
                </span>
                <svg
                  aria-hidden="true"
                  class="svg-inline--fa fa-chevron-down fa-w-14 margin-left-1"
                  data-icon="chevron-down"
                  data-prefix="fas"
                  focusable="false"
                  role="img"
                  viewBox="0 0 448 512"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M207.029 381.476L12.686 187.132c-9.373-9.373-9.373-24.569 0-33.941l22.667-22.667c9.357-9.357 24.522-9.375 33.901-.04L224 284.505l154.745-154.021c9.379-9.335 24.544-9.317 33.901.04l22.667 22.667c9.373 9.373 9.373 24.569 0 33.941L240.971 381.476c-9.373 9.372-24.569 9.372-33.942 0z"
                    fill="currentColor"
                  />
                </svg>
              </button>
            </th>
            <th
              aria-sort="none"
              colspan="1"
              role="columnheader"
              scope="col"
              style="min-width: 150px;"
            >
              <button
                aria-label="Toggle SortBy: updatedAt"
                class="usa-button usa-button--unstyled text-no-underline"
                style="cursor: pointer;"
                title="Toggle SortBy: updatedAt"
                type="button"
              >
                <span>
                  updatedAt
                </span>
                <svg
                  aria-hidden="true"
                  class="svg-inline--fa fa-chevron-down fa-w-14 margin-left-1"
                  data-icon="chevron-down"
                  data-prefix="fas"
                  focusable="false"
                  role="img"
                  viewBox="0 0 448 512"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M207.029 381.476L12.686 187.132c-9.373-9.373-9.373-24.569 0-33.941l22.667-22.667c9.357-9.357 24.522-9.375 33.901-.04L224 284.505l154.745-154.021c9.379-9.335 24.544-9.317 33.901.04l22.667 22.667c9.373 9.373 9.373 24.569 0 33.941L240.971 381.476c-9.373 9.372-24.569 9.372-33.942 0z"
                    fill="currentColor"
                  />
                </svg>
              </button>
            </th>
            <th
              aria-sort="none"
              colspan="1"
              role="columnheader"
              scope="col"
              style="min-width: 150px;"
            >
              <button
                aria-label="Toggle SortBy: icon"
                class="usa-button usa-button--unstyled text-no-underline"
                style="cursor: pointer;"
                title="Toggle SortBy: icon"
                type="button"
              >
                <span>
                  icon
                </span>
                <svg
                  aria-hidden="true"
                  class="svg-inline--fa fa-chevron-down fa-w-14 margin-left-1"
                  data-icon="chevron-down"
                  data-prefix="fas"
                  focusable="false"
                  role="img"
                  viewBox="0 0 448 512"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M207.029 381.476L12.686 187.132c-9.373-9.373-9.373-24.569 0-33.941l22.667-22.667c9.357-9.357 24.522-9.375 33.901-.04L224 284.505l154.745-154.021c9.379-9.335 24.544-9.317 33.901.04l22.667 22.667c9.373 9.373 9.373 24.569 0 33.941L240.971 381.476c-9.373 9.372-24.569 9.372-33.942 0z"
                    fill="currentColor"
                  />
                </svg>
              </button>
            </th>
          </tr>
        </thead>
        <tbody
          role="rowgroup"
        >
          <tr
            role="row"
          >
            <th
              role="cell"
              scope="row"
              style="background-color: rgb(240, 240, 240);"
            >
              <div>
                2
              </div>
            </th>
            <td
              role="cell"
            >
              <div
                class="text-secondary-vivid"
              >
                ! 1
              </div>
            </td>
            <td
              role="cell"
            >
              Banana
            </td>
            <td
              role="cell"
            >
              2021-11-11
            </td>
            <td
              role="cell"
            >
              fa-banana
            </td>
          </tr>
          <tr
            role="row"
          >
            <th
              role="cell"
              scope="row"
              style="background-color: rgb(240, 240, 240);"
            >
              <div>
                3
              </div>
            </th>
            <td
              role="cell"
            >
              <div
                class="text-secondary-vivid"
              >
                ! 2
              </div>
            </td>
            <td
              role="cell"
            >
              Chocolate
            </td>
            <td
              role="cell"
            >
              2020-11-11
            </td>
            <td
              role="cell"
            >
              fa-candy-bar
            </td>
          </tr>
          <tr
            role="row"
          >
            <th
              role="cell"
              scope="row"
              style="background-color: rgb(240, 240, 240);"
            >
              <div>
                4
              </div>
            </th>
            <td
              role="cell"
            >
              <div
                class="text-secondary-vivid"
              >
                ! 3
              </div>
            </td>
            <td
              role="cell"
            >
              Vanilla
            </td>
            <td
              role="cell"
            >
              2019-11-11
            </td>
            <td
              role="cell"
            >
              fa-ice-cream
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
  <br />
  <div
    class="grid-col-12 margin-top-3 font-sans-md text-bold"
  >
    Format columns
  </div>
  <div
    class="grid-col-12 font-sans-md font-sans-md text-bold"
  >
    <div
      class="margin-top-4"
    >
      <span
        class="text-base"
      >
        Column: id
      </span>
      <div
        class="grid-col-4"
      >
        <div
          class="usa-form-group margin-top-2"
          role="contentinfo"
        >
          <label
            class="usa-label text-bold"
            for="id"
          >
            Data format
          </label>
          <div
            class="usa-hint"
            id="id-description"
          />
          <select
            aria-describedby="id-description"
            class="usa-select"
            id="id"
            name="dataType"
          >
            <option
              value=""
            >
              Select an option
            </option>
            <option
              value="Text"
            >
              Text
            </option>
            <option
              value="Number"
            >
              Number
            </option>
            <option
              value="Date"
            >
              Date
            </option>
          </select>
        </div>
      </div>
      <div
        class="grid-col-4"
      >
        <div
          class="usa-form-group margin-top-3"
          role="contentinfo"
        >
          <label
            class="usa-label text-bold"
            for="id"
          >
            Number format (optional)
          </label>
          <div
            class="usa-hint"
            id="id-description"
          />
          <select
            aria-describedby="id-description"
            class="usa-select"
            id="id"
            name="numberType"
          >
            <option
              value=""
            >
              Select an option
            </option>
            <option
              value="Percentage"
            >
              Percentage
            </option>
            <option
              value="Currency"
            >
              Currency
            </option>
            <option
              value="With thousands separators"
            >
              With thousands separators
            </option>
          </select>
        </div>
      </div>
      <br />
      <hr />
    </div>
    <div
      class="margin-top-4"
    >
      <span
        class="text-base"
      >
        Column: name
      </span>
      <div
        class="grid-col-4"
      >
        <div
          class="usa-form-group margin-top-2"
          role="contentinfo"
        >
          <label
            class="usa-label text-bold"
            for="name"
          >
            Data format
          </label>
          <div
            class="usa-hint"
            id="name-description"
          />
          <select
            aria-describedby="name-description"
            class="usa-select"
            id="name"
            name="dataType"
          >
            <option
              value=""
            >
              Select an option
            </option>
            <option
              value="Text"
            >
              Text
            </option>
            <option
              value="Number"
            >
              Number
            </option>
            <option
              value="Date"
            >
              Date
            </option>
          </select>
        </div>
      </div>
      <br />
      <hr />
    </div>
    <div
      class="margin-top-4"
    >
      <span
        class="text-base"
      >
        Column: updatedAt
      </span>
      <div
        class="grid-col-4"
      >
        <div
          class="usa-form-group margin-top-2"
          role="contentinfo"
        >
          <label
            class="usa-label text-bold"
            for="updatedAt"
          >
            Data format
          </label>
          <div
            class="usa-hint"
            id="updatedAt-description"
          />
          <select
            aria-describedby="updatedAt-description"
            class="usa-select"
            id="updatedAt"
            name="dataType"
          >
            <option
              value=""
            >
              Select an option
            </option>
            <option
              value="Text"
            >
              Text
            </option>
            <option
              value="Number"
            >
              Number
            </option>
            <option
              value="Date"
            >
              Date
            </option>
          </select>
        </div>
      </div>
      <br />
      <hr />
    </div>
    <div
      class="margin-top-4"
    >
      <span
        class="text-base"
      >
        Column: icon
      </span>
      <div
        class="grid-col-4"
      >
        <div
          class="usa-form-group margin-top-2"
          role="contentinfo"
        >
          <label
            class="usa-label text-bold"
            for="icon"
          >
            Data format
          </label>
          <div
            class="usa-hint"
            id="icon-description"
          />
          <select
            aria-describedby="icon-description"
            class="usa-select"
            id="icon"
            name="dataType"
          >
            <option
              value=""
            >
              Select an option
            </option>
            <option
              value="Text"
            >
              Text
            </option>
            <option
              value="Number"
            >
              Number
            </option>
            <option
              value="Date"
            >
              Date
            </option>
          </select>
        </div>
      </div>
      <br />
      <hr />
    </div>
  </div>
  <br />
  <button
    class="usa-button usa-button--outline margin-top-1"
    type="button"
  >
    Back
  </button>
  <button
    class="usa-button usa-button--base margin-top-1"
    type="button"
  >
    Continue
  </button>
  <button
    class="usa-button usa-button--unstyled text-base-dark hover:text-base-darker active:text-base-darkest margin-top-1"
    type="button"
  >
    Cancel
  </button>
</div>
`;
